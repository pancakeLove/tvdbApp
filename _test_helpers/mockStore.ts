import configureMockStore from 'redux-mock-store';
import createSagaMiddleware from 'redux-saga';

export const mockStore = (state: Object, tvdbApi: any, saga: any) => {
  const context = {
    tvdbApi,
  };
  const sagaMiddleware = createSagaMiddleware();
  sagaMiddleware.setContext(context);
  const mockStore = configureMockStore([sagaMiddleware]);
  const store = mockStore(state);
  sagaMiddleware.run(saga);
  return {...store};
};
