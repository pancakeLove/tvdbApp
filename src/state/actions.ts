import {createStandardAction} from 'typesafe-actions';
import {SearchResultItem, SeriesDetails} from '../api/tvdb/types';
import {GeneralError} from '../shared/types';

export enum TvDbActionTypes {
  LOGIN_REQUEST = 'LOGIN_REQUEST',
  LOGIN_SUCCESS = 'LOGIN_SUCCESS',
  LOGIN_FAILURE = 'LOGIN_FAILURE',
  SEARCH_REQUEST = 'SEARCH_REQUEST',
  SEARCH_SUCCESS = 'SEARCH_SUCCESS',
  DETAILS_REQUEST = 'DETAILS_REQUEST',
  DETAILS_SUCCESS = 'DETAILS_SUCCESS',
  API_ERROR = 'API_ERROR',
  IS_LOADING = 'IS_LOADING',
}

export const tvDbActions = {
  loginRequest: createStandardAction(TvDbActionTypes.LOGIN_REQUEST)<null>(),
  loginSuccess: createStandardAction(TvDbActionTypes.LOGIN_SUCCESS)<string>(),
  loginFailure: createStandardAction(TvDbActionTypes.LOGIN_FAILURE)(),
  searchRequest: createStandardAction(TvDbActionTypes.SEARCH_REQUEST)<string>(),
  searchSuccess: createStandardAction(TvDbActionTypes.SEARCH_SUCCESS)<
    SearchResultItem[]
  >(),
  detailsRequest: createStandardAction(
    TvDbActionTypes.DETAILS_REQUEST,
  )<number>(),
  detailsSuccess: createStandardAction(
    TvDbActionTypes.DETAILS_SUCCESS,
  )<SeriesDetails | null>(),
  apiError: createStandardAction(
    TvDbActionTypes.API_ERROR,
  )<GeneralError | null>(),
  isLoading: createStandardAction(TvDbActionTypes.IS_LOADING)<boolean>(),
};
