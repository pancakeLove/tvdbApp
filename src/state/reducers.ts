import {Reducer} from 'redux';
import {TvDbActionTypes} from './actions';
import {Action} from 'typesafe-actions';
import {SearchResultItem, SeriesDetails} from '../api/tvdb/types';
import {GeneralError} from '../shared/types';

export interface TvDbState {
  loginInit: boolean;
  token: string;
  loginError: boolean;
  searchResult?: SearchResultItem[];
  apiError?: GeneralError;
  isLoading: boolean;
  seriesDetails?: SeriesDetails;
}

const initialState = {
  loginInit: false,
  token: '',
  loginError: false,
  searchResult: [],
  apiError: {error: false, message: ''},
  isLoading: false,
};

export interface DispatchAction extends Action<TvDbActionTypes> {
  payload: Partial<TvDbState>;
}

export const tvDbReducer: Reducer<TvDbState, DispatchAction> = (
  state = initialState,
  action,
) => {
  switch (action?.type) {
    case TvDbActionTypes.LOGIN_REQUEST:
      return {...state, loginError: false, loginInit: true};
    case TvDbActionTypes.LOGIN_SUCCESS:
      return {...state, token: action.payload as string, isLoading: false};
    case TvDbActionTypes.LOGIN_FAILURE:
      return {
        ...state,
        token: '',
        loginError: true,
      };
    case TvDbActionTypes.SEARCH_SUCCESS:
      return {
        ...state,
        searchResult: action.payload as SearchResultItem[],
        isLoading: false,
      };
    case TvDbActionTypes.DETAILS_SUCCESS:
      return {
        ...state,
        seriesDetails: action.payload as SeriesDetails,
        isLoading: false,
      };
    case TvDbActionTypes.API_ERROR:
      return {
        ...state,
        apiError: action.payload as GeneralError,
        isLoading: false,
      };
    default:
      return state;
  }
};
