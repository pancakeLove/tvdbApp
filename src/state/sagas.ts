import {
  takeLatest,
  call,
  put,
  all,
  select,
  getContext,
} from 'redux-saga/effects';
import {tvDbActions, TvDbActionTypes} from './actions';
import {
  LoginResponse,
  SearchResponse,
  SeriesDetailsResponse,
} from '../api/tvdb/types';
import {AxiosError} from 'axios';
import {TvDbState} from './reducers';
import {PayloadAction} from 'typesafe-actions';

function* handleApiError(error: AxiosError<any>) {
  let message: string = '';
  if (error.response) {
    message = 'Server responded with non 2xx status: ' + error.response.status;
  } else if (error.request) {
    message = 'Timeout received, possible slow network';
  } else {
    message = error.message;
  }
  yield put(tvDbActions.apiError({error: true, message}));
}

export function* loginWatcher() {
  yield takeLatest(TvDbActionTypes.LOGIN_REQUEST, loginWorker);
}
export function* loginWorker() {
  yield put(tvDbActions.isLoading(true));
  const tvdbApi = yield getContext('tvdbApi');
  //console.log('tvdbApi from context:', tvdbApi);
  const response: LoginResponse = yield call(tvdbApi().fetchToken);
  if (response.error) {
    yield put({type: TvDbActionTypes.LOGIN_FAILURE});
    yield handleApiError(response.error);
    yield put(tvDbActions.isLoading(false));
  } else {
    yield put(tvDbActions.loginSuccess(response.token));
  }
}

export function* searchWatcher() {
  yield takeLatest(TvDbActionTypes.SEARCH_REQUEST, searchWorker);
}
function* searchWorker(action: PayloadAction<any, string>) {
  yield put(tvDbActions.isLoading(true));
  const tvdbApi = yield getContext('tvdbApi');
  const token = yield select((state: TvDbState) => state.token);
  const response: SearchResponse = yield call(
    tvdbApi().search,
    action.payload,
    token,
  );
  if (response.error) {
    yield handleApiError(response.error);
    yield put(tvDbActions.isLoading(false));
  } else {
    /* todo: should be some limitation to number of results, eg:
     yield put(tvDbActions.searchSuccess(response.data?.slice(0, 100)));
     */
    yield put(tvDbActions.searchSuccess(response.data));
  }
}

export function* fetchDetailsWatcher() {
  yield takeLatest(TvDbActionTypes.DETAILS_REQUEST, fetchDetailsWorker);
}
function* fetchDetailsWorker(action: PayloadAction<any, number>) {
  yield put(tvDbActions.isLoading(true));
  const tvdbApi = yield getContext('tvdbApi');
  const token = yield select((state: TvDbState) => state.token);
  const response: SeriesDetailsResponse = yield call(
    tvdbApi().details,
    action.payload,
    token,
  );
  if (response.error) {
    yield handleApiError(response.error);
    yield put(tvDbActions.isLoading(false));
  } else {
    yield put(tvDbActions.detailsSuccess(response.data));
  }
}

export function* tvDbSagas() {
  yield all([loginWatcher(), searchWatcher(), fetchDetailsWatcher()]);
}
