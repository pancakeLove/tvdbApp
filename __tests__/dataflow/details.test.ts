import {tvDbActions} from '../../src/state/actions';
import {tvDbSagas} from '../../src/state/sagas';
import {mockStore} from '../../_test_helpers/mockStore';
import {
  SeriesDetails,
  SeriesDetailsResponse,
} from '../../src/api/tvdb/types';
import {AxiosError} from 'axios';

describe('dataflow for fetching details', () => {
  it('should dispatch a successful request to fetch details', async () => {
    const detailsSuccessResponse: SeriesDetailsResponse = {
      data: {} as SeriesDetails,
      error: undefined,
    };
    const tvdbApi = () => ({
      details: jest.fn().mockReturnValue(detailsSuccessResponse),
    });
    const store = mockStore({}, tvdbApi, tvDbSagas);
    await store.dispatch(tvDbActions.detailsRequest(1));
    expect(store.getActions()).toMatchSnapshot();
  });
  it('should dispatch a failing request to fetch details', async () => {
    const detailsFailsResponse: SeriesDetailsResponse = {
      data: {} as SeriesDetails,
      error: {} as AxiosError,
    };
    const tvdbApi = () => ({
      search: jest.fn().mockReturnValue(detailsFailsResponse),
    });
    const store = mockStore({}, tvdbApi, tvDbSagas);
    await store.dispatch(tvDbActions.searchRequest('query'));
    expect(store.getActions()).toMatchSnapshot();
  });
});
