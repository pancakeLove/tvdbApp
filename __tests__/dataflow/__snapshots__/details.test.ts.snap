// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dataflow for fetching details should dispatch a failing request to fetch details 1`] = `
Array [
  Object {
    "meta": undefined,
    "payload": "query",
    "type": "SEARCH_REQUEST",
  },
  Object {
    "meta": undefined,
    "payload": true,
    "type": "IS_LOADING",
  },
  Object {
    "meta": undefined,
    "payload": Object {
      "error": true,
      "message": undefined,
    },
    "type": "API_ERROR",
  },
  Object {
    "meta": undefined,
    "payload": false,
    "type": "IS_LOADING",
  },
]
`;

exports[`dataflow for fetching details should dispatch a successful request to fetch details 1`] = `
Array [
  Object {
    "meta": undefined,
    "payload": 1,
    "type": "DETAILS_REQUEST",
  },
  Object {
    "meta": undefined,
    "payload": true,
    "type": "IS_LOADING",
  },
  Object {
    "meta": undefined,
    "payload": Object {},
    "type": "DETAILS_SUCCESS",
  },
]
`;
